shared-bindings/busio/I2C.c

```diff
@@ -228,7 +228,7 @@ STATIC mp_obj_t busio_i2c_readfrom_into(size_t n_args, const mp_obj_t *pos_args,
 }
 MP_DEFINE_CONST_FUN_OBJ_KW(busio_i2c_readfrom_into_obj, 3, busio_i2c_readfrom_into);
 
-//|     def writeto(self, address: int, buffer: bytearray, *, start: int = 0, end: int = None, stop: bool = True) -> Any:
+//|     def writeto(self, address: int, buffer: bytearray, *, start: int = 0, end: int = None, stop: bool = True) -> None:
```

Comment:
A lot of these `bytes` and `bytearray` types are really just checking that something obeys "buf" protocol.
The actual argument validation checks in C are:
```c
    mp_get_buffer_raise(buffer, &bufinfo, MP_BUFFER_WRITE);   // Buffer must be writable.
// or
    mp_get_buffer_raise(buffer, &bufinfo, MP_BUFFER_READ);    // Buffer need only be readable.
```
How to specify this in type annotations is still an open issue in Python-land:
https://github.com/python/typing/issues/593
https://bugs.python.org/issue27501

In [Adafruit_Bleak_bleio](https://github.com/adafruit/Adafruit_Blinka_bleio), I have been doing
```python
Buf = Union[bytes, bytearray, memoryview]
```
in the actual code.

I won't mark all the cases in the PR, but there are many.

Reply:
This might be worth its own discussion somewhere, maybe in #3108.

Reply:
Ok, I'll change them to that

Reply:
Aha, this two-week-old PR: https://github.com/python/typeshed/pull/4232/files

added these defs:
```python
ReadableBuffer = Union[bytes, bytearray, memoryview, array.array, mmap.mmap]
WriteableBuffer = Union[bytearray, memoryview, array.array, mmap.mmap]
```
We don't have `mmap`.

Reply:
So I think `ReadableBuffer` and `WriteableBuffer` would be the best choices for now

Reply:
Shouldn't:
```
ReadableBuffer = Union[bytes, bytearray, memoryview, array.array, mmap.mmap]
WriteableBuffer = Union[bytearray, memoryview, array.array, mmap.mmap]
```
Be:
```
ReadableBuffer = TypeVar['ReadableBuffer', bytes, bytearray, memoryview, array.array, mmap.mmap]
WriteableBuffer = TypeVar['WriteableBuffer', bytearray, memoryview, array.array, mmap.mmap]
```
To account for functions that accept and then return the same buffer? Like `AnyStr` is!

Reply:
@hlovatt what is the advantage of TypeVar over Union? We're all new to Python types.